diff --recursive --unified --new-file 1.21.1/fabric/build.gradle.kts 1.20.6/fabric/build.gradle.kts
--- 1.21.1/fabric/build.gradle.kts	2025-05-27 12:20:42.495680200 +0300
+++ 1.20.6/fabric/build.gradle.kts	2025-05-27 12:20:42.480953600 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-Fabric-1.21.1"
+base.archivesName = "IAS-Fabric-1.20.6"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.1-root")
-val shared = project(":1.21.1-root")
+evaluationDependsOn(":1.20.6-root")
+val shared = project(":1.20.6-root")
 
 repositories {
     mavenCentral()
@@ -59,13 +59,13 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1211)
+    minecraft(libs.minecraft.mc1206)
     mappings(loom.officialMojangMappings())
 
     // Fabric
     modImplementation(libs.fabric.loader)
-    modImplementation(libs.fabric.mc1211)
-    modImplementation(libs.modmenu.mc1211)
+    modImplementation(libs.fabric.mc1206)
+    modImplementation(libs.modmenu.mc1206)
 
     // Root
     compileOnly(shared)
@@ -102,7 +102,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-Fabric-1.21.1",
+            "Implementation-Title" to "IAS-Fabric-1.20.6",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu"
         )
diff --recursive --unified --new-file 1.21.1/fabric/gradle.properties 1.20.6/fabric/gradle.properties
--- 1.21.1/fabric/gradle.properties	2025-05-27 12:20:42.495680200 +0300
+++ 1.20.6/fabric/gradle.properties	2025-05-27 12:20:42.480953600 +0300
@@ -16,6 +16,6 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>
 
 # Mod
-display.type=Fabric/Quilt 1.21.1
-modrinth.versions=1.21.1
+display.type=Fabric/Quilt 1.20.6
+modrinth.versions=1.20.6
 modrinth.loaders=fabric,quilt
diff --recursive --unified --new-file 1.21.1/fabric/src/main/resources/fabric.mod.json 1.20.6/fabric/src/main/resources/fabric.mod.json
--- 1.21.1/fabric/src/main/resources/fabric.mod.json	2025-05-27 12:20:42.497797200 +0300
+++ 1.20.6/fabric/src/main/resources/fabric.mod.json	2025-05-27 12:20:42.481958200 +0300
@@ -34,7 +34,7 @@
     "ias.mixins.json"
   ],
   "depends": {
-    "minecraft": "1.21.1",
+    "minecraft": "1.20.6",
     "fabric-api": "*"
   },
   "recommends": {
diff --recursive --unified --new-file 1.21.1/fabric/src/main/resources/quilt.mod.json 1.20.6/fabric/src/main/resources/quilt.mod.json
--- 1.21.1/fabric/src/main/resources/quilt.mod.json	2025-05-27 12:20:42.497797200 +0300
+++ 1.20.6/fabric/src/main/resources/quilt.mod.json	2025-05-27 12:20:42.483461300 +0300
@@ -27,8 +27,8 @@
     "depends": [
       {
         "id": "minecraft",
-        "versions": "1.21.1",
-        "reason": "In-Game Account Switcher works only with Minecraft 1.21.1."
+        "versions": "1.20.6",
+        "reason": "In-Game Account Switcher works only with Minecraft 1.20.6."
       },
       {
         "id": "fabric-api",
diff --recursive --unified --new-file 1.21.1/forge/build.gradle.kts 1.20.6/forge/build.gradle.kts
--- 1.21.1/forge/build.gradle.kts	2025-05-27 12:20:42.498875700 +0300
+++ 1.20.6/forge/build.gradle.kts	2025-05-27 12:20:42.483461300 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-Forge-1.21.1"
+base.archivesName = "IAS-Forge-1.20.6"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.1-root")
-val shared = project(":1.21.1-root")
+evaluationDependsOn(":1.20.6-root")
+val shared = project(":1.20.6-root")
 
 loom {
     silentMojangMappingsLicense()
@@ -63,11 +63,11 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1211)
+    minecraft(libs.minecraft.mc1206)
     mappings(loom.officialMojangMappings())
 
     // Forge
-    forge(libs.forge.mc1211)
+    forge(libs.forge.mc1206)
 
     // Root
     compileOnly(shared)
@@ -104,7 +104,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-Forge-1.21.1",
+            "Implementation-Title" to "IAS-Forge-1.20.6",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu",
             "MixinConfigs" to "ias.mixins.json"
diff --recursive --unified --new-file 1.21.1/forge/gradle.properties 1.20.6/forge/gradle.properties
--- 1.21.1/forge/gradle.properties	2025-05-27 12:20:42.498875700 +0300
+++ 1.20.6/forge/gradle.properties	2025-05-27 12:20:42.483461300 +0300
@@ -19,6 +19,6 @@
 loom.platform=forge
 
 # Mod
-display.type=Forge 1.21.1
-modrinth.versions=1.21.1
+display.type=Forge 1.20.6
+modrinth.versions=1.20.6
 modrinth.loaders=forge
diff --recursive --unified --new-file 1.21.1/forge/src/main/resources/META-INF/mods.toml 1.20.6/forge/src/main/resources/META-INF/mods.toml
--- 1.21.1/forge/src/main/resources/META-INF/mods.toml	2025-05-27 12:20:42.501098500 +0300
+++ 1.20.6/forge/src/main/resources/META-INF/mods.toml	2025-05-27 12:20:42.485478800 +0300
@@ -26,7 +26,7 @@
 modId = "forge"
 type = "required"
 mandatory = true
-versionRange = "[51,53)"
+versionRange = "[50,51)"
 ordering = "NONE"
 side = "CLIENT"
 
@@ -34,7 +34,7 @@
 modId = "minecraft"
 type = "required"
 mandatory = true
-versionRange = "[1.21,1.21.1]"
+versionRange = "[1.20.5,1.20.6]"
 ordering = "NONE"
 side = "CLIENT"
 
diff --recursive --unified --new-file 1.21.1/forge/src/main/resources/pack.mcmeta 1.20.6/forge/src/main/resources/pack.mcmeta
--- 1.21.1/forge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.501098500 +0300
+++ 1.20.6/forge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.485478800 +0300
@@ -1,6 +1,6 @@
 {
   "pack": {
-    "pack_format": 34,
+    "pack_format": 32,
     "description": "In-Game Account Switcher"
   }
 }
diff --recursive --unified --new-file 1.21.1/neoforge/build.gradle.kts 1.20.6/neoforge/build.gradle.kts
--- 1.21.1/neoforge/build.gradle.kts	2025-05-27 12:20:42.501098500 +0300
+++ 1.20.6/neoforge/build.gradle.kts	2025-05-27 12:20:42.486510200 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-NeoForge-1.21.1"
+base.archivesName = "IAS-NeoForge-1.20.6"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.1-root")
-val shared = project(":1.21.1-root")
+evaluationDependsOn(":1.20.6-root")
+val shared = project(":1.20.6-root")
 
 loom {
     silentMojangMappingsLicense()
@@ -65,11 +65,11 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1211)
+    minecraft(libs.minecraft.mc1206)
     mappings(loom.officialMojangMappings())
 
     // NeoForge
-    neoForge(libs.neoforge.mc1211)
+    neoForge(libs.neoforge.mc1206)
 
     // Root
     compileOnly(shared)
@@ -106,7 +106,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-NeoForge-1.21.1",
+            "Implementation-Title" to "IAS-NeoForge-1.20.6",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu",
             "MixinConfigs" to "ias.mixins.json"
diff --recursive --unified --new-file 1.21.1/neoforge/gradle.properties 1.20.6/neoforge/gradle.properties
--- 1.21.1/neoforge/gradle.properties	2025-05-27 12:20:42.502101900 +0300
+++ 1.20.6/neoforge/gradle.properties	2025-05-27 12:20:42.486510200 +0300
@@ -19,6 +19,6 @@
 loom.platform=neoforge
 
 # Mod
-display.type=NeoForge 1.21.1
-modrinth.versions=1.21.1
+display.type=NeoForge 1.20.6
+modrinth.versions=1.20.6
 modrinth.loaders=neoforge
diff --recursive --unified --new-file 1.21.1/neoforge/src/main/resources/META-INF/neoforge.mods.toml 1.20.6/neoforge/src/main/resources/META-INF/neoforge.mods.toml
--- 1.21.1/neoforge/src/main/resources/META-INF/neoforge.mods.toml	2025-05-27 12:20:42.503605200 +0300
+++ 1.20.6/neoforge/src/main/resources/META-INF/neoforge.mods.toml	2025-05-27 12:20:42.488840000 +0300
@@ -29,7 +29,7 @@
 modId = "neoforge"
 type = "required"
 mandatory = true
-versionRange = "[21.0,21.2)"
+versionRange = "[20.5,20.7)"
 ordering = "NONE"
 side = "CLIENT"
 
@@ -37,7 +37,7 @@
 modId = "minecraft"
 type = "required"
 mandatory = true
-versionRange = "[1.21,1.21.1]"
+versionRange = "[1.20.5,1.20.6]"
 ordering = "NONE"
 side = "CLIENT"
 
diff --recursive --unified --new-file 1.21.1/neoforge/src/main/resources/pack.mcmeta 1.20.6/neoforge/src/main/resources/pack.mcmeta
--- 1.21.1/neoforge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.503605200 +0300
+++ 1.20.6/neoforge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.488840000 +0300
@@ -1,6 +1,6 @@
 {
   "pack": {
-    "pack_format": 34,
+    "pack_format": 32,
     "description": "In-Game Account Switcher"
   }
 }
diff --recursive --unified --new-file 1.21.1/root/build.gradle.kts 1.20.6/root/build.gradle.kts
--- 1.21.1/root/build.gradle.kts	2025-05-27 12:20:42.503605200 +0300
+++ 1.20.6/root/build.gradle.kts	2025-05-27 12:20:42.488840000 +0300
@@ -25,7 +25,7 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-1.21.1"
+base.archivesName = "IAS-1.20.6"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
 
 repositories {
@@ -38,12 +38,12 @@
 }
 
 dependencies {
-    // Annotations
+    // Annotations (Compile)
     compileOnlyApi(libs.jetbrains.annotations)
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1211)
+    minecraft(libs.minecraft.mc1206)
     mappings(loom.officialMojangMappings())
 
     // Mixin
@@ -78,7 +78,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-1.21.1",
+            "Implementation-Title" to "IAS-1.20.6",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu"
         )
diff --recursive --unified --new-file 1.21.1/root/src/main/java/ru/vidtu/ias/IASMinecraft.java 1.20.6/root/src/main/java/ru/vidtu/ias/IASMinecraft.java
--- 1.21.1/root/src/main/java/ru/vidtu/ias/IASMinecraft.java	2025-05-27 12:20:42.505787100 +0300
+++ 1.20.6/root/src/main/java/ru/vidtu/ias/IASMinecraft.java	2025-05-27 12:20:42.490030300 +0300
@@ -82,9 +82,9 @@
      * Button widget sprites.
      */
     private static final WidgetSprites BUTTON = new WidgetSprites(
-            ResourceLocation.fromNamespaceAndPath("ias", "button_plain"),
-            ResourceLocation.fromNamespaceAndPath("ias", "button_disabled"),
-            ResourceLocation.fromNamespaceAndPath("ias", "button_focus")
+            new ResourceLocation("ias", "button_plain"),
+            new ResourceLocation("ias", "button_disabled"),
+            new ResourceLocation("ias", "button_focus")
     );
 
     /**
diff --recursive --unified --new-file 1.21.1/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java 1.20.6/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java
--- 1.21.1/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java	2025-05-27 12:20:42.506867600 +0300
+++ 1.20.6/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java	2025-05-27 12:20:42.491069900 +0300
@@ -51,26 +51,26 @@
      * Up button sprites.
      */
     private static final WidgetSprites UP = new WidgetSprites(
-            ResourceLocation.fromNamespaceAndPath("ias", "up_plain"),
-            ResourceLocation.fromNamespaceAndPath("ias", "up_disabled"),
-            ResourceLocation.fromNamespaceAndPath("ias", "up_focus")
+            new ResourceLocation("ias", "up_plain"),
+            new ResourceLocation("ias", "up_disabled"),
+            new ResourceLocation("ias", "up_focus")
     );
 
     /**
      * Down button sprites.
      */
     private static final WidgetSprites DOWN = new WidgetSprites(
-            ResourceLocation.fromNamespaceAndPath("ias", "down_plain"),
-            ResourceLocation.fromNamespaceAndPath("ias", "down_disabled"),
-            ResourceLocation.fromNamespaceAndPath("ias", "down_focus")
+            new ResourceLocation("ias", "down_plain"),
+            new ResourceLocation("ias", "down_disabled"),
+            new ResourceLocation("ias", "down_focus")
     );
 
     /**
      * Warning sprites.
      */
     private static final WidgetSprites WARNING = new WidgetSprites(
-            ResourceLocation.fromNamespaceAndPath("ias", "warning_off"),
-            ResourceLocation.fromNamespaceAndPath("ias", "warning_on")
+            new ResourceLocation("ias", "warning_off"),
+            new ResourceLocation("ias", "warning_on")
     );
 
     /**
diff --recursive --unified --new-file 1.21.1/root/src/main/java/ru/vidtu/ias/screen/LoginPopupScreen.java 1.20.6/root/src/main/java/ru/vidtu/ias/screen/LoginPopupScreen.java
--- 1.21.1/root/src/main/java/ru/vidtu/ias/screen/LoginPopupScreen.java	2025-05-27 12:20:42.507955500 +0300
+++ 1.20.6/root/src/main/java/ru/vidtu/ias/screen/LoginPopupScreen.java	2025-05-27 12:20:42.493577600 +0300
@@ -248,17 +248,6 @@
 
                 // Wow, opacity. So fluent.
                 // For what purpose?
-                // Actually this code and comment will probably never be read
-                // by anyone, so here's one little fact that makes me explode right now:
-                // >>>
-                // I have already written this code, however, I do have one little fun thing,
-                // I have my SSDs on my table, and so I accidentally "unplugged" them with my elbow.
-                // Not only it didn't save "unsaved" RAM, not only it did corrupt the file,
-                // it corrupted the BTRFS. I'm now writing this code once again from Win11 at 4 AM in the morn,
-                // it's actually pretty light-ish outside (imgur.com/a/waAxTK1), but this won't be pushed
-                // for a while, particularly for a reason I'm reinstalling my distro tomorrow and hell I'll use EXT4.
-                // Good night, thank you for using this mod and reading the source. <3
-                // <<<
                 int opacityMask;
                 if (this.error < 1.0F) {
                     this.error = Math.min(this.error + delta * 0.1F, 1.0F);
diff --recursive --unified --new-file 1.21.1/root/src/main/resources/ias.mixins.json 1.20.6/root/src/main/resources/ias.mixins.json
--- 1.21.1/root/src/main/resources/ias.mixins.json	2025-05-27 12:20:42.509964300 +0300
+++ 1.20.6/root/src/main/resources/ias.mixins.json	2025-05-27 12:20:42.495680200 +0300
@@ -1,7 +1,7 @@
 {
   "required": true,
   "minVersion": "0.8",
-  "compatibilityLevel": "JAVA_21",
+  "compatibilityLevel": "JAVA_17",
   "package": "ru.vidtu.ias.mixins",
   "refmap": "ias.mixins.refmap.json",
   "client": [
