diff --recursive --unified --new-file 1.21.3/fabric/build.gradle.kts 1.21.1/fabric/build.gradle.kts
--- 1.21.3/fabric/build.gradle.kts	2025-05-27 12:20:42.510967700 +0300
+++ 1.21.1/fabric/build.gradle.kts	2025-05-27 12:20:42.495680200 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-Fabric-1.21.3"
+base.archivesName = "IAS-Fabric-1.21.1"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.3-root")
-val shared = project(":1.21.3-root")
+evaluationDependsOn(":1.21.1-root")
+val shared = project(":1.21.1-root")
 
 repositories {
     mavenCentral()
@@ -59,13 +59,13 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1213)
+    minecraft(libs.minecraft.mc1211)
     mappings(loom.officialMojangMappings())
 
     // Fabric
     modImplementation(libs.fabric.loader)
-    modImplementation(libs.fabric.mc1213)
-    modImplementation(libs.modmenu.mc1213)
+    modImplementation(libs.fabric.mc1211)
+    modImplementation(libs.modmenu.mc1211)
 
     // Root
     compileOnly(shared)
@@ -102,7 +102,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-Fabric-1.21.3",
+            "Implementation-Title" to "IAS-Fabric-1.21.1",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu"
         )
diff --recursive --unified --new-file 1.21.3/fabric/gradle.properties 1.21.1/fabric/gradle.properties
--- 1.21.3/fabric/gradle.properties	2025-05-27 12:20:42.510967700 +0300
+++ 1.21.1/fabric/gradle.properties	2025-05-27 12:20:42.495680200 +0300
@@ -16,6 +16,6 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>
 
 # Mod
-display.type=Fabric/Quilt 1.21.3
-modrinth.versions=1.21.3
+display.type=Fabric/Quilt 1.21.1
+modrinth.versions=1.21.1
 modrinth.loaders=fabric,quilt
diff --recursive --unified --new-file 1.21.3/fabric/src/main/resources/fabric.mod.json 1.21.1/fabric/src/main/resources/fabric.mod.json
--- 1.21.3/fabric/src/main/resources/fabric.mod.json	2025-05-27 12:20:42.513054200 +0300
+++ 1.21.1/fabric/src/main/resources/fabric.mod.json	2025-05-27 12:20:42.497797200 +0300
@@ -34,7 +34,7 @@
     "ias.mixins.json"
   ],
   "depends": {
-    "minecraft": "1.21.3",
+    "minecraft": "1.21.1",
     "fabric-api": "*"
   },
   "recommends": {
diff --recursive --unified --new-file 1.21.3/fabric/src/main/resources/quilt.mod.json 1.21.1/fabric/src/main/resources/quilt.mod.json
--- 1.21.3/fabric/src/main/resources/quilt.mod.json	2025-05-27 12:20:42.513054200 +0300
+++ 1.21.1/fabric/src/main/resources/quilt.mod.json	2025-05-27 12:20:42.497797200 +0300
@@ -27,8 +27,8 @@
     "depends": [
       {
         "id": "minecraft",
-        "versions": "1.21.3",
-        "reason": "In-Game Account Switcher works only with Minecraft1.21.3."
+        "versions": "1.21.1",
+        "reason": "In-Game Account Switcher works only with Minecraft 1.21.1."
       },
       {
         "id": "fabric-api",
diff --recursive --unified --new-file 1.21.3/forge/build.gradle.kts 1.21.1/forge/build.gradle.kts
--- 1.21.3/forge/build.gradle.kts	2025-05-27 12:20:42.514052600 +0300
+++ 1.21.1/forge/build.gradle.kts	2025-05-27 12:20:42.498875700 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-Forge-1.21.3"
+base.archivesName = "IAS-Forge-1.21.1"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.3-root")
-val shared = project(":1.21.3-root")
+evaluationDependsOn(":1.21.1-root")
+val shared = project(":1.21.1-root")
 
 loom {
     silentMojangMappingsLicense()
@@ -63,11 +63,11 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1213)
+    minecraft(libs.minecraft.mc1211)
     mappings(loom.officialMojangMappings())
 
     // Forge
-    forge(libs.forge.mc1213)
+    forge(libs.forge.mc1211)
 
     // Root
     compileOnly(shared)
@@ -104,7 +104,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-Forge-1.21.3",
+            "Implementation-Title" to "IAS-Forge-1.21.1",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu",
             "MixinConfigs" to "ias.mixins.json"
diff --recursive --unified --new-file 1.21.3/forge/gradle.properties 1.21.1/forge/gradle.properties
--- 1.21.3/forge/gradle.properties	2025-05-27 12:20:42.514052600 +0300
+++ 1.21.1/forge/gradle.properties	2025-05-27 12:20:42.498875700 +0300
@@ -19,6 +19,6 @@
 loom.platform=forge
 
 # Mod
-display.type=Forge 1.21.3
-modrinth.versions=1.21.3
+display.type=Forge 1.21.1
+modrinth.versions=1.21.1
 modrinth.loaders=forge
diff --recursive --unified --new-file 1.21.3/forge/src/main/resources/META-INF/mods.toml 1.21.1/forge/src/main/resources/META-INF/mods.toml
--- 1.21.3/forge/src/main/resources/META-INF/mods.toml	2025-05-27 12:20:42.515200800 +0300
+++ 1.21.1/forge/src/main/resources/META-INF/mods.toml	2025-05-27 12:20:42.501098500 +0300
@@ -26,7 +26,7 @@
 modId = "forge"
 type = "required"
 mandatory = true
-versionRange = "[53,54)"
+versionRange = "[51,53)"
 ordering = "NONE"
 side = "CLIENT"
 
@@ -34,7 +34,7 @@
 modId = "minecraft"
 type = "required"
 mandatory = true
-versionRange = "[1.21.2,1.21.3]"
+versionRange = "[1.21,1.21.1]"
 ordering = "NONE"
 side = "CLIENT"
 
diff --recursive --unified --new-file 1.21.3/forge/src/main/resources/pack.mcmeta 1.21.1/forge/src/main/resources/pack.mcmeta
--- 1.21.3/forge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.516199900 +0300
+++ 1.21.1/forge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.501098500 +0300
@@ -1,6 +1,6 @@
 {
   "pack": {
-    "pack_format": 42,
+    "pack_format": 34,
     "description": "In-Game Account Switcher"
   }
 }
diff --recursive --unified --new-file 1.21.3/neoforge/build.gradle.kts 1.21.1/neoforge/build.gradle.kts
--- 1.21.3/neoforge/build.gradle.kts	2025-05-27 12:20:42.516199900 +0300
+++ 1.21.1/neoforge/build.gradle.kts	2025-05-27 12:20:42.501098500 +0300
@@ -25,10 +25,10 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-NeoForge-1.21.3"
+base.archivesName = "IAS-NeoForge-1.21.1"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
-evaluationDependsOn(":1.21.3-root")
-val shared = project(":1.21.3-root")
+evaluationDependsOn(":1.21.1-root")
+val shared = project(":1.21.1-root")
 
 loom {
     silentMojangMappingsLicense()
@@ -65,11 +65,11 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1213)
+    minecraft(libs.minecraft.mc1211)
     mappings(loom.officialMojangMappings())
 
     // NeoForge
-    neoForge(libs.neoforge.mc1213)
+    neoForge(libs.neoforge.mc1211)
 
     // Root
     compileOnly(shared)
@@ -106,7 +106,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-NeoForge-1.21.3",
+            "Implementation-Title" to "IAS-NeoForge-1.21.1",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu",
             "MixinConfigs" to "ias.mixins.json"
diff --recursive --unified --new-file 1.21.3/neoforge/gradle.properties 1.21.1/neoforge/gradle.properties
--- 1.21.3/neoforge/gradle.properties	2025-05-27 12:20:42.516199900 +0300
+++ 1.21.1/neoforge/gradle.properties	2025-05-27 12:20:42.502101900 +0300
@@ -19,6 +19,6 @@
 loom.platform=neoforge
 
 # Mod
-display.type=NeoForge 1.21.3
-modrinth.versions=1.21.3
+display.type=NeoForge 1.21.1
+modrinth.versions=1.21.1
 modrinth.loaders=neoforge
diff --recursive --unified --new-file 1.21.3/neoforge/src/main/resources/META-INF/neoforge.mods.toml 1.21.1/neoforge/src/main/resources/META-INF/neoforge.mods.toml
--- 1.21.3/neoforge/src/main/resources/META-INF/neoforge.mods.toml	2025-05-27 12:20:42.518257400 +0300
+++ 1.21.1/neoforge/src/main/resources/META-INF/neoforge.mods.toml	2025-05-27 12:20:42.503605200 +0300
@@ -29,7 +29,7 @@
 modId = "neoforge"
 type = "required"
 mandatory = true
-versionRange = "[21.2,21.4)"
+versionRange = "[21.0,21.2)"
 ordering = "NONE"
 side = "CLIENT"
 
@@ -37,7 +37,7 @@
 modId = "minecraft"
 type = "required"
 mandatory = true
-versionRange = "[1.21.2,1.21.3]"
+versionRange = "[1.21,1.21.1]"
 ordering = "NONE"
 side = "CLIENT"
 
diff --recursive --unified --new-file 1.21.3/neoforge/src/main/resources/pack.mcmeta 1.21.1/neoforge/src/main/resources/pack.mcmeta
--- 1.21.3/neoforge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.518257400 +0300
+++ 1.21.1/neoforge/src/main/resources/pack.mcmeta	2025-05-27 12:20:42.503605200 +0300
@@ -1,6 +1,6 @@
 {
   "pack": {
-    "pack_format": 42,
+    "pack_format": 34,
     "description": "In-Game Account Switcher"
   }
 }
diff --recursive --unified --new-file 1.21.3/root/build.gradle.kts 1.21.1/root/build.gradle.kts
--- 1.21.3/root/build.gradle.kts	2025-05-27 12:20:42.519262000 +0300
+++ 1.21.1/root/build.gradle.kts	2025-05-27 12:20:42.503605200 +0300
@@ -25,7 +25,7 @@
 java.targetCompatibility = JavaVersion.VERSION_21
 java.toolchain.languageVersion = JavaLanguageVersion.of(21)
 group = "ru.vidtu.ias"
-base.archivesName = "IAS-1.21.3"
+base.archivesName = "IAS-1.21.1"
 description = "This mod allows you to change your logged in account in-game, without restarting Minecraft."
 
 repositories {
@@ -43,7 +43,7 @@
     compileOnlyApi(libs.error.prone.annotations)
 
     // Minecraft (Provided)
-    minecraft(libs.minecraft.mc1213)
+    minecraft(libs.minecraft.mc1211)
     mappings(loom.officialMojangMappings())
 
     // Mixin
@@ -78,7 +78,7 @@
             "Specification-Title" to "In-Game Account Switcher",
             "Specification-Version" to version,
             "Specification-Vendor" to "VidTu",
-            "Implementation-Title" to "IAS-1.21.3",
+            "Implementation-Title" to "IAS-1.21.1",
             "Implementation-Version" to version,
             "Implementation-Vendor" to "VidTu"
         )
diff --recursive --unified --new-file 1.21.3/root/src/main/java/ru/vidtu/ias/IASMinecraft.java 1.21.1/root/src/main/java/ru/vidtu/ias/IASMinecraft.java
--- 1.21.3/root/src/main/java/ru/vidtu/ias/IASMinecraft.java	2025-05-27 12:20:42.520363800 +0300
+++ 1.21.1/root/src/main/java/ru/vidtu/ias/IASMinecraft.java	2025-05-27 12:20:42.505787100 +0300
@@ -33,7 +33,7 @@
 import net.minecraft.client.gui.components.WidgetSprites;
 import net.minecraft.client.gui.components.events.GuiEventListener;
 import net.minecraft.client.gui.components.toasts.SystemToast;
-import net.minecraft.client.gui.components.toasts.ToastManager;
+import net.minecraft.client.gui.components.toasts.ToastComponent;
 import net.minecraft.client.gui.layouts.LayoutElement;
 import net.minecraft.client.gui.screens.ConnectScreen;
 import net.minecraft.client.gui.screens.Screen;
@@ -263,7 +263,7 @@
         }
 
         // Warn about invalid names.
-        if (!IASConfig.nickWarns || !(screen instanceof ConnectScreen) || minecraft.getToastManager().getToast(SystemToast.class, NICK_WARN) != null) return;
+        if (!IASConfig.nickWarns || !(screen instanceof ConnectScreen) || minecraft.getToasts().getToast(SystemToast.class, NICK_WARN) != null) return;
         User user = minecraft.getUser();
         // Mods break non-nullness.
         //noinspection ConstantValue
@@ -272,8 +272,8 @@
         if (key == null) return;
 
         // Display the toast.
-        ToastManager manager = minecraft.getToastManager();
-        manager.addToast(SystemToast.multiline(minecraft, NICK_WARN, Component.literal("In-Game Account Switcher"), Component.translatable(key, name)));
+        ToastComponent toasts = minecraft.getToasts();
+        toasts.addToast(SystemToast.multiline(minecraft, NICK_WARN, Component.literal("In-Game Account Switcher"), Component.translatable(key, name)));
     }
 
     /**
diff --recursive --unified --new-file 1.21.3/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java 1.21.1/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java
--- 1.21.3/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java	2025-05-27 12:20:42.521376600 +0300
+++ 1.21.1/root/src/main/java/ru/vidtu/ias/screen/AccountEntry.java	2025-05-27 12:20:42.506867600 +0300
@@ -29,7 +29,6 @@
 import net.minecraft.client.gui.components.WidgetSprites;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.client.gui.screens.inventory.tooltip.DefaultTooltipPositioner;
-import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.resources.PlayerSkin;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
@@ -158,7 +157,7 @@
         // Render warning if insecure.
         if (this.account.insecure()) {
             boolean warning = (System.nanoTime() / 1_000_000_000L) % 2L == 0;
-            graphics.blitSprite(RenderType::guiTextured, warning ? WARNING.enabled() : WARNING.enabledFocused(), x - 6, y - 1, 2, 10);
+            graphics.blitSprite(warning ? WARNING.enabled() : WARNING.enabledFocused(), x - 6, y - 1, 2, 10);
             if (mouseX >= x - 10 && mouseX <= x && mouseY >= y && mouseY <= y + height) {
                 this.list.screen().setTooltipForNextRenderPass(Tooltip.create(Component.translatable("ias.accounts.tip.insecure")), DefaultTooltipPositioner.INSTANCE, true);
             }
@@ -176,7 +175,7 @@
             } else {
                 upTexture = UP.enabled();
             }
-            graphics.blitSprite(RenderType::guiTextured, upTexture, upX, y, 11, 7);
+            graphics.blitSprite(upTexture, upX, y, 11, 7);
 
             // Render down widget.
             ResourceLocation downTexture;
@@ -188,7 +187,7 @@
             } else {
                 downTexture = DOWN.enabled();
             }
-            graphics.blitSprite(RenderType::guiTextured, downTexture, downX, y, 11, 7);
+            graphics.blitSprite(downTexture, downX, y, 11, 7);
         }
     }
 
